wordpress-mobile/release-toolkit
# Add Android AVD / emulator actions

Adds a helper and multiple actions to help us install Android system-images, create AVDs dynamically, and boot and shutdown Android emulators.

These are used by our `screenshots` lane in https://github.com/wordpress-mobile/WordPress-Android/pull/17096 to automate capture of raw screenshots.

> EDIT: Reworked to improve many error cases (e.g. `sdkmanager` installed via `brew` but `emulator` installed by Android Studio, leading to multiple SDK roots; or giving hints for the `PANIC: Broken AVD` message; etc.) and to support dynamic emulator TCP ports (i.e. non-hardcoded `emulator-5554` serial) to support running multiple emulators in parallel 💪 

## To Test

### Base Scenario

 - Switch to the `tooling/screenshots` branch in WPAndroid (from https://github.com/wordpress-mobile/WordPress-Android/pull/17096)
 - Run `bundle install` to make sure you point the `fastlane/Pluginfile` and `release-toolkit` in that WPAndroid branch to this toolkit's `screenshots/emulator-helper` branch
 - Run `bundle exec fastlane screenshots app:wordpress locale:zh-CN` from the WPAndroid repo
 - Ensure that the screenshot automation finishes and worked (it can take a couple of minutes†…), in particular:
   - Check that the logs mention calling the `android_create_avd` action, installing the system image and creating the AVD successfully, for each case (phone, tablet)
   - Check that the logs mention calling the `android_launch_emulator` action, booting the emulator then waiting for it to finish booting, for each case (phone, tablet)
   - Check that the lane ended up creating the raw screenshots in Chinese for both phone and tablet device sizes successfully.
   - Once the full lane completes and the screenshot UI tests have finished running, check that the raw screenshots images have been generated and pulled from the emulator's sdcard, and are present in the HTML summary file.

> **Note**: _† For some reason it appears that sometimes there's a long pause between when the unit test has started and the app has launched… and when the first click of the UI test scenario finally starts. I'm still not sure what could explain those delays (they are not consistent and are hard to debug), so for the time just don't be surprised that it could sometime be between 2 to 8 minutes between when the app's first screen shows in the emulator and when the UI test finally starts its clickity-clackity work._ 🙃 

### Variants / configurations

Please consider testing all these different setups, to ensure all my error handling of possible scenarios encountered on different user's machines or CI hosts will be covered

 - 💻  Test on both an Intel and an M1 machine
    - especially to validate that the installation of the appropriate Android System Image by `sdkmanager` installs the appropriate system image depending on the architecture of the machine you run those lanes from.
 - 🔧  Test with different ways of installing the Android Command Line Tools
    - Uninstall the command line tools like `sdkmanager` and `avdmanager`, then re-run the `screnshots` lane and confirm that the action errors, with a suggestion to install them
    - Install the Android command line tools [using `brew`](https://formulae.brew.sh/cask/android-commandlinetools) — while you keep `emulator` installed by Android Studio and reachable from terminal (either via your `$PATH` or by having `$ANDROID_SDK_ROOT` defined). Then re-run the `screenshots` lane and confirm that the action errors, suggesting that you might have two different Android SDK locations (brew and AS) and that `sdkmanager` and `emulator` might not be installed in the same one. Then `brew uninstall` the Android command line tools once done.
    - Finally, re-install the Command Line Tools via the Android Studio's "SDK Manager" window, and test the lane again. That time it should succeed.
 - 🏎️  Test to run multiple emulators in parallel
    - Start any AVD from Android Studio that is neither `Pixel_3_API_28` nor `Nexus_9_API_28`. The goal here is to have the TCP port `5554` — which is usually the default one / first picked by `adb` to connect to an emulator — already in use.
    - In one terminal window, run `bundle exec fastlane screenshots app:wordpress locale:ar device:phone`
    - While the first command is still running, run `bundle exec fastlane screenshots app:wordpress locale:ar device:tenInch` in another terminal window
    - Verify that each command found the correct serial (e.g. `emulator-5556` for the phone one and `emulator-5558` for the tablet one — since `emulator-5554` was likely the one already used by the dummy AVD you launched from AS in step one)
